### PR TITLE
StringArray comparisions return BooleanArray

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/pandas-dev/pandas/issues/29556

cc @jorisvandenbossche

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -59,6 +59,37 @@
     rxor,
 )
 
+# -----------------------------------------------------------------------------
+# constants
+ARITHMETIC_BINOPS: Set[str] = {
```
Comment (TomAugspurger) [inline pandas/core/ops/__init__.py]: @jbrockmendel am I reinventing any wheels here? This seems like something useful enough that it would have been done before.

File: doc/source/user_guide/text.rst [hunk 2]
```diff
@@ -94,6 +94,10 @@ l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
 2. Some string methods, like :meth:`Series.str.decode` are not available
    on ``StringArray`` because ``StringArray`` only holds strings, not
    bytes.
+3. In comparision operations, :class:`StringArray` and ``Series`` backed
+   by a ``StringArray`` will return an object with :class:`BooleanDtype`,
+   rather than a ``bool`` or ``object`` dtype object, depending on whether
```
Comment (jorisvandenbossche) [inline doc/source/user_guide/text.rst]: Currently, I don't think you get object dtype (None or NaN compares to False, so it still is a boolean result)

File: pandas/core/arrays/string_.py [hunk 3]
```diff
@@ -148,6 +152,13 @@ class StringArray(PandasArray):
     Traceback (most recent call last):
     ...
     ValueError: StringArray requires an object-dtype ndarray of strings.
+
+    For comparision methods, this returns a :class:`pandas.BooleanArray`
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py]:
```suggestion
    For comparision methods, this returns a :class:`pandas.arrays.BooleanArray`
```

File: pandas/tests/arrays/string_/test_string.py [hunk 4]
```diff
@@ -154,6 +154,39 @@ def test_add_frame():
...
 
 
+def test_comparison_methods_scalar(all_compare_operators):
+    op_name = all_compare_operators
+
+    a = pd.array(["a", None, "c"], dtype="string")
+    other = "a"
+    result = getattr(a, op_name)(other)
+    expected = np.array([getattr(item, op_name)(other) for item in a], dtype=object)
+    expected[1] = None
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/string_/test_string.py]: Why is this needed? Does the op with NA not give NA?

**Comment by TomAugspurger** [comment]
Thanks! Fixed all those I think.

Tangentially related: we'll need a docs section discussing the behavior of `pd.NA` and arrays using `pd.NA` in comparison operations (propagate rather than unequal). I think it makes sense to combine `user_guide/integer_na.rst` and `user_guide/boolean.rst` into a single "nullable types" document which discusses all of this? What do you think?

**Comment by jorisvandenbossche** [comment]
> I think it makes sense to combine user_guide/integer_na.rst and user_guide/boolean.rst into a single "nullable types" document which discusses all of this? What do you think?

I was floating the idea in my head (but didn't make an issue or so yet) to create a "data types" directory and move the integer and boolean files there (so an extra nesting in the left sidebar). And then the main "data types" page could have a short intro to extension dtypes, pointing to those pages. 
But, if the content is relatively limited, it might also make sense to just combine both on a single page for now.

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: i dont think so, no

File: pandas/core/ops/__init__.py [hunk 5]
```diff
@@ -59,6 +59,37 @@
...
+    "mod",
+    "floordiv",
+    "truediv",
+    "divmod",
+    "radd",
+    "rsub",
+    "rmul",
+    "rpow",
+    "rmod",
+    "rfloordiv",
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: is there a reason these are not fixtures?

**Review by jreback — CHANGES_REQUESTED** [review]
I don't think you are really testing this. I think you actually need to take some of our existing indexing tests, then paramterize over the existing object based strings and then StringArray, then test indexing. I suspect a fair amount of things will break.

File: pandas/core/ops/__init__.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/ops/__init__.py]: > is there a reason these are not fixtures?

This is used in *code*, not in tests

**Comment by jorisvandenbossche** [comment]
> I don't think you are really testing this. I think you actually need to take some of our existing indexing tests,

This does not yet add indexing, this only changes the comparison operations. Indexing is one of the next steps on the list in https://github.com/pandas-dev/pandas/issues/29556

**Comment by jreback** [comment]
> > I don't think you are really testing this. I think you actually need to take some of our existing indexing tests,
> 
> This does not yet add indexing, this only changes the comparison operations. Indexing is one of the next steps on the list in #29556

ok that's fair.

**Comment by TomAugspurger** [comment]
Indexing seems independent right? BooleanArray comparisons already return BooleanArray, and AFAIK those aren't tested in indexing yet.

I'm happy to write those tests / work on the implementation, but it feels like a separate piece of work.

**Comment by jorisvandenbossche** [comment]
Yes, let's keep it separate

**Comment by jreback** [comment]
> Indexing seems independent right? BooleanArray comparisons already return BooleanArray, and AFAIK those aren't tested in indexing yet.
> 
> I'm happy to write those tests / work on the implementation, but it feels like a separate piece of work.

they are separate until sometries to assign a BooleanArray to a column and use it. We have *nothing* for this, certainly do in a separate PR. What I am suggesting is a single test to validate that this raises (for now)

**Comment by TomAugspurger** [comment]
Do we know what the defined behavior is yet? In https://github.com/pandas-dev/pandas/issues/28778 it sounds like we're contemplating either raising or propagating.

File: doc/source/user_guide/text.rst [hunk 6]
```diff
@@ -94,7 +94,12 @@ l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
 2. Some string methods, like :meth:`Series.str.decode` are not available
    on ``StringArray`` because ``StringArray`` only holds strings, not
    bytes.
-
+3. In comparision operations, :class:`arrays.StringArray` and ``Series`` backed
+   by a ``StringArray`` will return an object with :class:`arrays.BooleanDtype`,
```
Comment (jorisvandenbossche) [inline doc/source/user_guide/text.rst]:
```suggestion
   by a ``StringArray`` will return an object with :class:`BooleanDtype`,
```

**Review by jorisvandenbossche — APPROVED** [review]

File: doc/source/user_guide/text.rst [hunk 7]
```diff
@@ -94,7 +94,12 @@ l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
...
+   rather than a ``bool`` dtype object, depending on whether
+   there are missing values. Missing values in a ``StringArray`` will propagate
```
Comment (jreback) [inline doc/source/user_guide/text.rst]: is there a reason we are not *always* returning a BooleanArray, or am I mis-reading this? this would be *very* confusing if this is indeed the case.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/user_guide/text.rst (hunk 7 above)
Comment (jorisvandenbossche) [reply doc/source/user_guide/text.rst]: I think it is stated confusingly. When the dtype is "string" (not object), comparisons always return BooleanDtype. 
Tom, I think you can removed the ", depending on whether there are missing values"

Comment (TomAugspurger) [reply doc/source/user_guide/text.rst]: Fixed.

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (WillAyd) [inline pandas/core/ops/__init__.py]: Just as FYI if you create a container with literal values mypy will infer that it is a `Set[str]` for you, so not required to specify. Have to specify if the container is empty on instantiation

**Review by WillAyd — COMMENTED** [review]
generally lgtm

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/ops/__init__.py]: Good to know!

**Comment by TomAugspurger** [comment]
This should be good to go.

**Comment by jorisvandenbossche** [comment]
Thanks!
